### PR TITLE
Enable set sequence nr

### PR DESCRIPTION
**Pull request by YLGH**
Summary:
In some cases (especially those involving collective calls) - we would want to always kick off a collective call first before running going down another path.

For  example:

```
tbe lookup -> a2a ->
                     overarch
dense ------------->
```

if the forward code is written as
a2a_out = a2a
dense = dense_net
out = overarch(a2a_out, dense)
out.backward()

The current default is running backwards in the opposite order the forward is called. However, there is no data dependency between a2a and dense, so in reality either of them could be run first. We would like the a2a to run first because it provides optimal (on average) overlap.

Changing the seq_nr of a2a_out to something large enough would allow autograd engine to kick it off first.

Test Plan: Tests incoming

Differential Revision: D51445261

### CODE REVIEW

**Review by ezyang — APPROVED** [review]
this LGTM, not sure if Alban you want to have the final word here though, if so just bounce my review

File: test/test_set_sequence_nr.py
Comment (albanD) [inline test/test_set_sequence_nr.py:1]: Please move this to `test/autograd/test_seq_nr.py` and import it at the bottom of `test/test_autograd.py`.
Feel free to just add it to `test/test_autograd.py` given that it is very small.

File: test/run_test.py
Comment (albanD) [inline test/run_test.py:327]: Why does this need to be in the serial list? It should be ok to run it anytime?

File: torch/csrc/autograd/python_cpp_function.cpp
Comment (albanD) [inline torch/csrc/autograd/python_cpp_function.cpp:208]:
```suggestion
  HANDLE_TH_ERRORS
  auto& fn = *((THPCppFunction*)self)->cdata;
  fn.set_sequence_nr(THPUtils_unpackUInt64(sequence_nr));
  Py_RETURN_NONE;
  END_HANDLE_TH_ERRORS
```

**Review by albanD — COMMENTED** [review]
Small details but mostly good.

File: test/run_test.py
Comment (YLGH) [reply test/run_test.py:327]: oh whoops I mostly had no idea how to write a test and just templated off of test_fake_tensor

File: torch/csrc/autograd/python_cpp_function.cpp
Comment (YLGH) [reply torch/csrc/autograd/python_cpp_function.cpp:208]: I didn't see any of the other methods in cpp_function having this (but I did have it in python_function.cpp)

Comment (albanD) [reply torch/csrc/autograd/python_cpp_function.cpp:208]: because most of them cannot raise c++ errors and handle python errors by hand (for historical reasons, these macro might not have existed when this code was written).
THPUtils_unpackUInt64 will throw a nice c++ error on error which you need to handle here.

Comment (YLGH) [reply torch/csrc/autograd/python_cpp_function.cpp:208]: ahh gotcha, thanks

File: test/test_autograd.py
Comment (albanD) [inline test/test_autograd.py:3]: I guess all the import shuffling is auto-formatting?
We usually don't do these unless they're from the lint to avoid shuffling back and forth depending on dev environments.

**Review by albanD — APPROVED** [review]
SGTM

File: test/test_autograd.py
Comment (YLGH) [reply test/test_autograd.py:3]: ah whoops - i'll let lint runner check to see if this is ok, otherwise i'll make it original format

Comment (YLGH) [reply test/test_autograd.py]: reverted formatting changes

**Review by albanD — APPROVED** [review]

**Pull request closed without merging**
